common/darwin/Classes/FlutterRTCDesktopCapturer.m

```diff
@@ -277,25 +301,149 @@ -(RTCDesktopSource *)getSourceById:(NSString *)sourceId {
+                                            binaryMessenger:messenger];
+        [_eventChannel setStreamHandler:self];
+    }
+}
+
+#pragma mark - FlutterStreamHandler methods
+
+- (FlutterError* _Nullable)onCancelWithArguments:(id _Nullable)arguments {
+    _eventSink = nil;
+    return nil;
+}
+
+- (FlutterError* _Nullable)onListenWithArguments:(id _Nullable)arguments
+                                       eventSink:(nonnull FlutterEventSink)sink {
+    _eventSink = sink;
+    return nil;
+}
+
+
+#pragma mark - RTCDesktopMediaListDelegate delegate
+
+- (void)didDesktopSourceAdded:(RTC_OBJC_TYPE(RTCDesktopSource) *)source {
+    //NSLog(@"didDesktopSourceAdded: %@, id %@", source.name, source.sourceId);
+    _captureSources = [_captureSources arrayByAddingObject:source];
+    if(_eventSink) {
+        NSImage *image = [source UpdateThumbnail];
+        NSData *data = [[NSData alloc] init];
+        if(image != nil) {
+            NSImage *resizedImg = [self resizeImage:image forSize:NSMakeSize(140, 140)];
+            data = [resizedImg TIFFRepresentation];
+        }
+        _eventSink(@{
+            @"event": @"desktopSourceAdded",
+            @"id": source.sourceId,
+            @"name": source.name,
+            @"thumbnailSize": @{@"width": @0, @"height": @0},
+            @"type": source.sourceType == RTCDesktopSourceTypeScreen? @"screen" : @"window",
+            @"thumbnail": data
+        });
+    }
+}
+
+- (void)didDesktopSourceRemoved:(RTC_OBJC_TYPE(RTCDesktopSource) *) source {
+   //NSLog(@"didDesktopSourceRemoved: %@, id %@", source.name, source.sourceId);
+    _captureSources = [_captureSources arrayByRemovingObject:source];
+    if(_eventSink) {
+        _eventSink(@{
+            @"event": @"desktopSourceRemoved",
+            @"id": source.sourceId,
+        });
+    }
+}
+
+- (void)didDesktopSourceNameChanged:(RTC_OBJC_TYPE(RTCDesktopSource) *) source {
+    //NSLog(@"didDesktopSourceNameChanged: %@, id %@", source.name, source.sourceId);
+    if(_eventSink) {
+        _eventSink(@{
+            @"event": @"desktopSourceNameChanged",
+            @"id": source.sourceId,
+            @"name": source.name,
+        });
+    }
+}
+
+- (void)didDesktopSourceThumbnailChanged:(RTC_OBJC_TYPE(RTCDesktopSource) *) source {
+    //NSLog(@"didDesktopSourceThumbnailChanged: %@, id %@", source.name, source.sourceId);
+    if(_eventSink) {
+        NSImage *resizedImg = [self resizeImage:[source thumbnail] forSize:NSMakeSize(140, 140)];
+        NSData *data = [resizedImg TIFFRepresentation];
+        _eventSink(@{
+            @"event": @"desktopSourceThumbnailChanged",
+            @"id": source.sourceId,
+            @"thumbnail": data 
+        });
+    }
+}
+
+#pragma mark - RTCDesktopCapturerDelegate delegate
+
+-(void)didSourceCaptureStart:(RTCDesktopCapturer *) capturer {
+    NSLog(@"didSourceCaptureStart");
+    _interruptSourcesRefresh = YES;
```

Comment:
Can you explain why this is needed? Does source capturing interfere with the updating of the sources list?

Reply:
Actually I'm going to delete the use of timer in native to refresh the list tomorrow (only `getSources`, `updateSources` two methods), so these codes will be removed